Added GLBufferAttribute

raub opened the pull request:
If present, would delegate VBO creation to the user-supplied function. See more info on this in #11883. The default behavior is untouched.

raub committed:
Added onCreateCallback for WebGLAttributes #11883

pailhead commented:
More hooks like this!!

raub committed:
Revert "Added onCreateCallback for WebGLAttributes #11882"

This reverts commit 152c85f1bdcb787725e04bf9a94dc1997aef3bca.

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

raub committed:
Added new buffer attribute class

raub commented:
Now a new `DynamicBufferAttribute` class exists. Example: **webgl_buffergeometry_points_dynamicbufferattribute.html**.

`new DynamicBufferAttribute(...)` parameters:
* gl - renderer.context,
* buffer - an actual VBO,
* type - e.g. gl.FLOAT,
* itemSize - vector size, e.g. 3 for xyz position,
* count - vertex count.

mrdoob commented:
Do you mind renaming it to `GLBufferAttribute`? `BufferAttribute` has a `dynamic` property (which sets `gl.DYNAMIC_DRAW) and I could see people getting confused.

raub commented:
I see your point, but all of them (buffer types) are of GL kind... `GLBufferAttribute` is not a perfect name either. Could it be something else? Like
* `ManualBufferAttribute`
* `ExternalBufferAttribute`
* `SharedBufferAttribute`
* (other suggestions welcome)

I think prefix `GL` tells really nothing about what is special in this new attribute class, so it should be the last resort name of choise.

raub committed:
dynamic buffer attr : fixed significant update overhead

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

raub committed:
rename to ExternalBufferAttribute

raub commented:
@mrdoob, since no answer from you, I've decided to take some action and finally rename the class to `ExternalBufferAttribute`. Is this fine? What else is to be done before the merge could hapen?

mrdoob commented:
I still prefer `GLBufferAttribute`. We're planning to also add `GLLine` for people that want to use raw `gl.LINE` instead of the new line with linewidth support. This name also hints the developer that no other renderers will be able to handle these classes.

Sorry I couldn't elaborate in time for r92, moving to r93.

raub commented:
Ok, will do soon

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

raub committed:
renamed to GLBufferAttribute, fixed lint errors

raub committed:
WebGLAttributes use get/set method for GLBufferAttibute

raub committed:
fix lint error

raub commented:
Done, `GLBufferAttribute`. Seems to work fine with the latest dev branch.

WestLangley commented:
Nice demo. :-)

Can you please: add your new example to `examples/files.js`,  update the docs with your new class, and check your formatting. https://zz85.github.io/mrdoobapproves/

Mugen87 commented on examples/webgl_buffergeometry_points_glbufferattribute.html:
```diff
@@ -0,0 +1,195 @@
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<title>three.js webgl - buffergeometry [particles]</title>
```
Maybe renaming the title to a more descriptive name: `three.js webgl - buffergeometry - custom VBOs`.

raub replied:
Ok, see bbc5638

raub replied:
Done

raub committed:
added GLBufferAttribute to docs

raub committed:
changed GLBufferAttribute example title

raub committed:
fixed missing space in GLBufferAttribute example

raub committed:
fixed info title in GLBufferAttribute example

raub commented:
@WestLangley
- Added the doc for `GLBufferAttribute` class. And listed it in the index.
- Added the example to the examples page.
- Could not find any more linting issues.

raub commented:
@Mugen87 could you resolve the review? if there is something else to do, then please tell me

raub committed:
Merge branch 'master' of https://github.com/mrdoob/three.js into dev

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

mrdoob commented on src/core/BufferGeometry.js:
```diff
@@ -597,6 +597,17 @@ BufferGeometry.prototype = Object.assign( Object.create( EventDispatcher.prototy
 
 		var position = this.attributes.position;
 
+		if ( position && position.isGLBufferAttribute ) {
```
How about instructing the user that the have to set `mesh.frustumCulled` to `false` instead?

mrdoob replied:
We kind of do that already sometimes when using `SkinnedMesh` or when running gpu particle simulations.

WestLangley replied:
>How about instructing the user that the have to set mesh.frustumCulled to false instead?

...or, alternatively, set the bounding sphere manually...

raub replied:
Hmm, I didn't know about `mesh.frustumCulled`, this is very useful to me, thanks.

I think here it's better to check if the mesh is not culled or if it has some manual bbox/sphere. Because we don't pull the vertices from GPU to calc the bbox, but user might still need it. The warning is to be shown when both culling enabled and no bbox given. I'll try to implement this soon.

raub replied:
I've simply added the warnings in f8f0b41. Now the example is set up to use either custom bounding sphere or no culling. I've checked that it shows the warning if I use neither of those in the example.

raub committed:
Add culling warnings for GLBufferAttribute, update the example

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

raub commented:
@mrdoob Hey, is there anything else to be done before this gets merged?

Not only I wait for it because I hoped to switch my dependent module to use the main threejs package (rather then my fork), but also I'm getting manual merges here and there as my code is out of the common codebase.

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

raub committed:
Fix GLBufferAttribute example

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

raub committed:
Add  parameter to the constructor

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

mrschofield commented:
Not to be a pain, but is there any chance this could be merged?

Is there any outstanding work that is required?

raub committed:
Update GLBufferAttribute

raub committed:
Merge branch 'master' of https://github.com/mrdoob/three.js into dev

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

raub committed:
Add example screenshot

raub committed:
Fix e2e test

mrdoob commented:
Okay, planning on merging this next week. Finally... 😅

mrdoob commented on docs/list.js:
```diff
@@ -90,6 +90,7 @@ var list = {
 				"EventDispatcher": "api/en/core/EventDispatcher",
 				"Face3": "api/en/core/Face3",
 				"Geometry": "api/en/core/Geometry",
+				"GLBufferAttribute": "api/{lan}/core/GLBufferAttribute",
```
This should be:

```js
"GLBufferAttribute": "api/en/core/GLBufferAttribute",
```

raub replied:
done

raub committed:
Fix docs list entry

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

Mugen87 commented:
Can you please save a copy of `docs/api/en/core/GLBufferAttribute.html` at `docs/api/zh/core/GLBufferAttribute.html` and also add then Chinese entry it `docs/list.js`?

raub committed:
Add zh doc entry

Mugen87 commented:
The TS declaration file for `GLBufferAttribute` should be something like:
```ts
export class GLBufferAttribute {

	constructor( gl: WebGLRenderingContext, buffer: WebGLBuffer, type: number, itemSize: number, count: number, normalized: boolean );

	buffer: WebGLBuffer;
	type: number;
	itemSize: number;
	elementSize: number;
	count: number;
	normalized: boolean;
	version: number;
	sizes: [ [ number, number ] ];

	readonly isGLBufferAttribute: true;

	set needsUpdate( value: boolean );

	setBuffer( buffer: WebGLBuffer ): this;
	setType( type: number ): this;
	setItemSize( itemSize: number ): this;
	setCount( count: number ): this;

}
```

Mugen87 commented on src/core/GLBufferAttribute.js:
```diff
@@ -0,0 +1,96 @@
+
+function GLBufferAttribute( gl, buffer, type, itemSize, count, normalized ) {
+
+	this.sizes = [
```
Is this property supposed to be private? Is so I would remove it from the public API.

raub replied:
The object stored in `this.sizes` is used by methods below. I also need `gl` to build this object. This is why I build it at the construction time, when `gl` is by definition available. Being a 'public' member is just a side-effect of the above. What would you suggest?

Mugen87 replied:
It seems you need `sizes ` only to determine `elementSize`, right? Check my comment below. I think it's better if the user defines `elementSize`.

Mugen87 commented on src/core/GLBufferAttribute.js:
```diff
@@ -0,0 +1,96 @@
+	this.buffer = buffer;
+	this.type = type;
+	this.itemSize = itemSize;
+	this.elementSize = this.sizes[ type ];
```
Is `elementSize` also intended to be private?

raub replied:
It is intended to be public. The idea is as follows: you know the type of your data, you pass it, the `elementSize` becomes available, carrying the size-in-bytes value for you.

It is not necessary. We could let the app do all the calculations and leave here only what's necessary for three.js to handle the attribute.

raub replied:
Ah wait, I found it here: src\renderers\webgl\WebGLAttributes.js
```
			buffers.set( attribute, {
				buffer: attribute.buffer,
				type: attribute.type,
				bytesPerElement: attribute.elementSize,
				version: attribute.version
			} );
```
so basically it is there for three.js to get the byte size, because there is no `Array` instance to do so (`bytesPerElement: array.BYTES_PER_ELEMENT,`).

Mugen87 replied:
Um, have you considered to let the user set this value? I mean `GLBufferAttribute` is obviously a low-level API not necessarily intended for beginners. Devs who use it should also know the correct bytes per element.

I would to put as less logic as possible into the ctor and let instead the user defined such information. In this way, you probably don't even need a WebGL rendering context in `GLBufferAttribute` which I find a bit strange anyway.

raub replied:
The logic behind this is that passing both `type` and `elementSize` to the constructor is redundant, because the `type` always implies the only possible `elementSize`. So it kind of creates a room for an error to occur when you ask the user to pass both. So I decided to calculate it on the ctor side.

But in general it would be fine either way. If the ctor takes `elementSize`, we get rid of all the logic and the contraversial `this.sizes` member. Also no need for `gl` then. Do you approve of the latter scenario?

Mugen87 replied:
I would say yes. Keeping `GLBufferAttribute` simple and let the app handle this bit sound like the better design for me. Especially since we can assume the user knows the relationship between `type` and `elementSize` if this class is going to be used.

The final decision is up to @mrdoob though 😉

raub replied:
Any updates on that matter? @Mugen87 @mrdoob

Mugen87 replied:
Let's wait for @mrdoob's feedback.

raub commented:
`sizes: [ [ number, number ] ];` I think should be `sizes: { [number]: number };` something. Unsure if this is a fine piece of TS syntax though. How can we validate it?

Mugen87 commented:
If you are unsure about this, use `any` for the moment^^.

zlf2002171 commented:
unsubscribe

2020-05-18 

zlf2002171 



发件人：Luis Blanco <notifications@github.com>
发送时间：2020-04-29 22:53
主题：Re: [mrdoob/three.js] Added GLBufferAttribute (#13196)
收件人："mrdoob/three.js"<three.js@noreply.github.com>
抄送："Subscribed"<subscribed@noreply.github.com>

@raub commented on this pull request.



In src/core/GLBufferAttribute.js:
> +
+		accum[ current[ 0 ] ] = current[ 1 ];
+		return accum;
+
+	}, {} );
+
+	if ( ! this.sizes[ type ] ) {
+
+		throw new TypeError( 'THREE.GLBufferAttribute: unsupported GL data type.' );
+
+	}
+
+	this.buffer = buffer;
+	this.type = type;
+	this.itemSize = itemSize;
+	this.elementSize = this.sizes[ type ];

Any updates on that matter? @Mugen87 @mrdoob
—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

pvaananen commented:
Sorry for bothering you but I'm really looking forward to this addition. So the current status is that the PR is waiting for @mrdoob's feedback?

Regarding that TypeScript question on Apr 27, may I suggest declaring an array of number pairs with `sizes:Array<[number, number]>` or `sizes:[number, number][]`.
```

mrdoob commented:
Ah, too close to release now... Lets see if we can merge this at the beginning on the next dev cycle.

Mugen87 commented:
@mrdoob Before merging, can you please review https://github.com/mrdoob/three.js/pull/13196/files#r416037718? I'm not sure about the current public interface of `GLBufferAttribute`.

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

raub committed:
Update GLBufferAttribute

raub commented:
1. Removed "gl" constructor param.
2. Removed "normalized" constructor param.
3. Added "elementSize" constructor param.
4. Removed "sizes" property.
5. Updated the TS class declaration.
6. Updated the docs.
7. Updated the example to fit new interface.
8. Updated the example logic for e2e compliance.
9. Debugged and fixed an issue when buffers don't change under VAO mode. (pls review!)
10. Generated a screenshot and passed e2e.

raub committed:
Update screenshot for GLBufferAttribute example

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

Mugen87 commented:
Thanks @raub! This looks much better to me now!

raub committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into dev

mrdoob commented:
I think you can delete `webgl_buffergeometry_points_glbufferattribute.png`.

mrdoob commented:
I'll merge and clean up.

mrdoob commented:
Thanks!

Mugen87 commented on src/renderers/webgl/WebGLBindingStates.js:
```diff
@@ -176,6 +176,8 @@
 			const cachedAttribute = cachedAttributes[ key ];
 			const geometryAttribute = geometryAttributes[ key ];
 
+			if ( geometryAttribute.isGLBufferAttribute ) return true;
```
Can you please explain why you have added this line? When removed, the respective demo does not seem to work correctly anymore. 

I don't think it's correct to always force a call of `setupVertexAttributes()` on each frame.

raub replied:
I don't exactly understand what this part of the system is supposed to do. But my guess is:

- if there is no VAO, bind appropriate buffers
- if there is a VAO and the attributes were not changed, just switch to that VAO
- if attributes changed, update the VAO, i.e. switch to it and bind the updated buffers

That all sounds good in theory. But when I noticed my demo was broken, I started debugging and ended up here. So basically I set `needsUpdate` to my attribute and expect the engine to do the job of updating whatever was changed (the buffer). That was not the case here.

For some reason, `cachedAttribute` always contained an already-updated version of my attribute. So the VAO never got updated (hence the demo was broken). I double-checked this fact by adding some custom properties directly into the buffer, like `buffer.name=111`.

So I decided to opt-out from this optimization with my attribute type. I expect that line 179 effectively makes VAO unused for `GLBufferAttribute`. That is fine with me. Although if you understand what is going wrong and how to make it work with this optimization, that would also be great.

Mugen87 replied:
The problem is that the current implementation is bad for performance since a lot of unnecessary state changes are done. 

We definitely have to revisit this issue otherwise I would vote to revert `GLBufferAttribute` before the next release.

raub replied:
The current implementation is vastly superior than having no implementation. And if I'm right, for `webgl1` people the further discussion makes no sense, since they always get exactly this level of performance. So there is a considerable group of people who either get this feature or not, regardless of VAO optimizations.

Now to optimizations. I would gladly fixed the VAO issue, given it was at all possible for you (or anybody) to explan why the **cached** attribute is always the updated attribute itself. And if it is the correct behavior, how it can be possibly fixed.

If you want, I might go on and report this behavior as a bug in a separate issue. Because now it is my understanding, that when I set `needsUpdate` on my attribute, the cache should contain the old version. If I have just updated the attribute, and the "cached" version does already contain what I have updated, **it is a bug in WebGLBindingStates** unless proven otherwise.

Mugen87 replied:
This has nothing to do with WebGL 1 or 2. VAO is available for most WebGL 1 users. Besides, even without VAO the buffer attribute bindings should only be done if necessary. Not if just the contents of a buffer changes.

I think the problem in your demo is this line:
```js
posAttr.buffer = ( posAttr.buffer === pos ) ? pos2 : pos;
```
You should exchange the entire attribute and not just the buffer. Or you keep the buffer the same and only change the contents.

Mugen87 replied:
This pattern works without the hack in `WebGLBindingStates`:
```js
var posAttr1 = new THREE.GLBufferAttribute( pos, gl.FLOAT, 3, 4, particles );
var posAttr2 = new THREE.GLBufferAttribute( pos2, gl.FLOAT, 3, 4, particles );
geometry.setAttribute( 'position', posAttr1 );

setInterval( function () {

    var attr = geometry.getAttribute( 'position' );

    geometry.setAttribute( 'position', ( attr === posAttr1 ) ? posAttr2 : posAttr1 );
 
}, 2000 );
```
